README.md
[![Rust](https://github.com/muaddibus/bevy_webcam_facial/actions/workflows/rust.yml/badge.svg)](https://github.com/muaddibus/bevy_webcam_facial/actions/workflows/rust.yml)

# About bevy_webcam_facial

Plugin for [Bevy](https://bevyengine.org/) game engine. Captures webcam image, finds face and provides all available data (face rectangle coordinates, face probability) to Bevy game engine via events for further use in Bevy game engine.

## Features

* Webcam capture using [camera_capture](https://github.com/oli-obk/camera_capture)
* Face position recognition using [rustface](https://github.com/atomashpolskiy/rustface)
* Realtime and lightweight [SeetaFace Detection model](https://github.com/seetaface/SeetaFaceEngine/tree/master/FaceDetection/)
* Runs in separate Bevy AsyncTaskpool task without blocking
* 2 data smoothing/denoising filters

## Plans
- [ ] MacOSX webcam support
- [ ] Several AI face recognition models to choose by default (simple frame, with face features like eyes/nose/mouth, full face mesh recognition, emotion detection...)

## Supported Platforms

- [x] Linux
- [ ] MacOSX
- [x] Windows

## Available for use in Bevy:

### Plugin config

Several parameters when including in `.add_plugins` or use default camera and settings `.add_plugins(WebcamFacialPlugin::default())`:
```rust
.add_plugins(WebcamFacialPlugin {
    config_webcam_device: 0,
    config_webcam_width: 640,
    config_webcam_height: 480,
    config_webcam_framerate: 15,
    config_webcam_autostart: true,
    config_filter_type: SmoothingFilterType::LowPass(0.1),
    config_filter_length: 10,
})
```
Parameters: 
* Webcamera device number (0-first default) ex.0,1,2...
    * Linux: Number get appended to `/dev/video{number}`
    * Windows: Device number
* Width of frame: 640
* Width of frame: 480
* Frames per second: 15
* Start capturing and sending events instantly after plugin activation: true/false (can be enabled/disabled anytime at runtime via `ResMut<WebcamFacialController>`)
* Smoothing filter for coordinates (currently: MeanMedian, LowPass(f32), NoFilter)
* From how many frames take data for smoothing 5-10 optimal (more frames - less noisy data, but slower response)

### Resources:
Enable/disable webcam capture and recognition from Bevy via mutable resource `ResMut<WebcamFacialController>`
```rust
pub struct WebcamFacialController {
...
    pub control: bool,
...
}
```
### Event with captured data
```rust
<Event>WebcamFacialDataEvent
```
### Data struct returned via Event
```rust
pub struct WebcamFacialData {
    pub center_x: f32,
    pub center_y: f32,
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
    pub score: f32,
}
```
Coordinates are mapped as floating point number in range of -50.0 .. 50.0, camera resolution doesn't matter
* [center_x) Face center point x coordinate
* (center_y) Face center point y coordinate
* (x) Face rectangle frame x coordinate
* (y) Face rectangle frame y coordinate
* (width) Face rectangle frame width
* (height) Face rectangle frame height
* (score) Probability of a detected object being a true face 0-30..


## Some ideas and use cases of data comming from plugin:
* Controlling game object transformations (transform, rotate, scale)
* Object control (car driving, player movement...)
* Background scene movement in 2D games or background scene movement in 3D top/side view games for better depth perception or 'looking around'
* Camera FPS like movement
* Rotation around scenes, player or other objects
* Zooming in scenes (map zoom, scene zoom, sniper zoom...)
* Scaring horror games to pop beasts on detected face closeup
* Your imagination...

*Note: Use some interpolation for transforms for smoother transforms like "bevy_easings" or "bevy_mod_interp"


## Examples
Three examples are provided in [examples] folder:
(under construction)
- [x] [object_mover](examples/object_mover.rs) - simplest example to move object using raw unfiltered/noisy data
- [x] [camera_control](examples/camera_control.rs) - control bevy camera view using filtered data
- [ ] [neck_trainer](examples/neck_trainer.rs) - train you neck :) most complex example with filtered data + bone animation and skin

Unchecked - not finished

## Versions

| bevy | bevy_webcam_facial  |
|  ---:|                 ---:|
| 0.11 | 0.1.4               |
| 0.11 | 0.1.3               |
| 0.11 | 0.1.2               |
| 0.11 | 0.1.1               |

[![Bevy tracking](https://img.shields.io/badge/Bevy%20tracking-released%20version-lightblue)](https://github.com/bevyengine/bevy/blob/main/docs/plugins_guidelines.md#main-branch-tracking)

## Reference Material

The following were used for coding of plugin:

* [Bevy The Book](https://bevyengine.org/learn/book/)
* [Unofficial Bevy Cheat Book](https://bevy-cheatbook.github.io/0)
* [Bevy Third Party Plugin Guidelines](https://github.com/bevyengine/bevy/blob/main/docs/plugins_guidelines.md)
* [rustface](https://github.com/atomashpolskiy/rustface)

Additional interesting sources for future research:

* [Tensorflow rust](https://github.com/tensorflow/rust/) 
* [Face Detection with Tensorflow Rust](https://cetra3.github.io/blog/face-detection-with-tensorflow-rust/) (Was succesfully tested earlier)
* [Tensorflow model implementing the mtcnn face detector](https://github.com/blaueck/tf-mtcnn/)
* https://github.com/cetra3/mtcnn/
* [tfjs-model web demo of face mesh](https://storage.googleapis.com/tfjs-models/demos/face-landmarks-detection/index.html?model=mediapipe_face_mesh)
* [Awesome-Rust-MachineLearning Docs/Projects](https://github.com/vaaaaanquish/Awesome-Rust-MachineLearning/)

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any
additional terms or conditions.


CHANGELOG.md
## 0.1.4 - 2023-08-07
### Added
Default settings for plugin:
```rust
            config_webcam_device: 0,
            config_webcam_width: 640,
            config_webcam_height: 480,
            config_webcam_framerate: 15,
            config_webcam_autostart: true,
            config_filter_type: SmoothingFilterType::LowPass(0.1),
            config_filter_length: 10,
```
Added more informative exception handling
### Changed
Plugin now uses [camera_capture] instead of [rscam]. Should bring Windows support.
config_webcam_device is now u32 type. 0 - default camera, or device number (Linux: number is added to "/dev/video{}", Windows: device number)
### Removed
Converter from yuyv not needed anymore

## 0.1.3 - 2023-08-05
### Added
Filters to smooth noise recognition data. 2 new fields in plugin configuration:
```rust
.add_plugins(WebcamFacialPlugin {
    config_webcam_device: "/dev/video0".to_string(),
    config_webcam_width: 640,
    config_webcam_height: 480,
    config_webcam_framerate: 33,
    config_webcam_autostart: true,
    config_filter_type: SmoothingFilterType::LowPass(0.1),
    config_filter_length: 10,
})
```
Added default-features = false to Cargo.toml

### Changed
<Event>WebcamFacialDataEvent now returns float coordinate values in range -50.0 .. 50.0, counted from center of frame
```rust
pub struct WebcamFacialData {
    pub center_x: f32,
    pub center_y: f32,
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
    pub score: f32,
}
```

Updated second example (Still needs nice scene)

## 0.1.0 - 2023-07-24
### Added


src/filter.rs
use crate::WebcamFacialData;

pub struct WebcamFacialDataFiltered(Vec<WebcamFacialData>, u32, SmoothingFilterType);

#[allow(unused)]
#[derive(Clone, Copy)]
pub enum SmoothingFilterType {
    MeanMedian,
    LowPass(f32), // LowPass filter with alpha parameter
    NoFilter,
}

impl WebcamFacialDataFiltered {
    pub fn new(length: u32, filter_type: SmoothingFilterType) -> Self {
        Self(Vec::new(), length, filter_type)
    }

    #[allow(unused)]
    pub fn push(&mut self, data: WebcamFacialData) {
        if self.0.len() >= self.1 as usize {
            self.0.remove(0); // Remove the oldest element
        }
        self.0.push(data);
    }

    #[allow(unused)]
    pub fn get(&mut self) -> WebcamFacialData {
        if self.0.len() == 0 {
            self.push(WebcamFacialData::default());
        }
        match self.2 {
            SmoothingFilterType::MeanMedian => return self.mean_median_filter(),
            SmoothingFilterType::LowPass(alpha) => return self.low_pass_filter(alpha),
            SmoothingFilterType::NoFilter => {
                return WebcamFacialData {
                    center_x: self.0[self.0.len() - 1].center_x,
                    center_y: self.0[self.0.len() - 1].center_y,
                    x: self.0[self.0.len() - 1].x,
                    y: self.0[self.0.len() - 1].y,
                    width: self.0[self.0.len() - 1].width,
                    height: self.0[self.0.len() - 1].height,
                    score: self.0[self.0.len() - 1].score,
                }
            }
        }
    }

    #[allow(unused)]
    fn low_pass_filter(&self, alpha: f32) -> WebcamFacialData {
        let mut filtered_data = WebcamFacialData::default();
        for data in &self.0 {
            filtered_data.center_x += alpha * (data.center_x - filtered_data.center_x);
            filtered_data.center_y += alpha * (data.center_y - filtered_data.center_y);
            filtered_data.x += alpha * (data.x - filtered_data.x);
            filtered_data.y += alpha * (data.y - filtered_data.y);
            filtered_data.width += alpha * (data.width - filtered_data.width);
            filtered_data.height += alpha * (data.height - filtered_data.height);
            filtered_data.score += alpha * (data.score - filtered_data.score);
        }

        filtered_data
    }

    #[allow(unused)]
    fn mean_median_filter(&self) -> WebcamFacialData {
        let num_elements = self.0.len();

        let mut center_x_sum = 0.0;
        let mut center_y_sum = 0.0;
        let mut x_sum = 0.0;
        let mut y_sum = 0.0;
        let mut width_sum = 0.0;
        let mut height_sum = 0.0;
        let mut score_sum = 0.0;

        for data in &self.0 {
            center_x_sum += data.center_x;
            center_y_sum += data.center_y;
            x_sum += data.x;
            y_sum += data.y;
            width_sum += data.width;
            height_sum += data.height;
            score_sum += data.score;
        }

        WebcamFacialData {
            center_x: center_x_sum / num_elements as f32,
            center_y: center_y_sum / num_elements as f32,
            x: x_sum / num_elements as f32,
            y: y_sum / num_elements as f32,
            width: width_sum / num_elements as f32,
            height: height_sum / num_elements as f32,
            score: score_sum / num_elements as f32,
        }
    }
}


examples/neck_trainer.rs
use bevy::prelude::*;
use bevy_scene_hook::{HookPlugin, HookedSceneBundle, SceneHook};

use bevy_webcam_facial::*;

fn main() {
    App::new()
        .insert_resource(Average::default())
        .insert_resource(ClearColor(Color::WHITE))
        .insert_resource(AmbientLight {
            color: Color::WHITE,
            brightness: 1.0,
        })
        .add_plugins(DefaultPlugins)
        .add_plugins(WebcamFacialPlugin {
            config_webcam_device: 0,
            config_webcam_width: 640,
            config_webcam_height: 480,
            config_webcam_framerate: 33,
            config_webcam_autostart: true,
            config_filter_length: 15,
            config_filter_type: SmoothingFilterType::LowPass(0.1),
        })
        .add_plugins(HookPlugin)
        .add_systems(Startup, load_scene)
        .add_systems(Update, (keyboard_animation_control, bone_move))
        .run();
}

#[derive(Resource, Default)]
struct Average {
    x: f32,
    y: f32,
}

#[derive(Debug)]
struct Animations(Vec<Handle<AnimationClip>>);

#[derive(Component)]
struct HeadBone;

/// set up 3D scene
fn load_scene(mut cmds: Commands, asset_server: Res<AssetServer>) {
    cmds.spawn(HookedSceneBundle {
        scene: SceneBundle {
            scene: asset_server.load("rooster.gltf#Scene0"),
            ..default()
        },
        hook: SceneHook::new(|entity, cmds| {
            match entity.get::<Name>().map(|t| t.as_str()) {
                Some("Headas") => cmds.insert(HeadBone),
                _ => cmds,
            };
        }),
    });
}

fn keyboard_animation_control(
    keyboard_input: Res<Input<KeyCode>>,
    mut animation_player: Query<&mut AnimationPlayer>,
) {
    if let Ok(mut player) = animation_player.get_single_mut() {
        if keyboard_input.just_pressed(KeyCode::Space) {
            if player.is_paused() {
                player.resume();
            } else {
                player.pause();
            }
        }

        if keyboard_input.just_pressed(KeyCode::Up) {}
    }
}
fn bone_move(
    mut head: Query<&mut Transform, With<HeadBone>>,
    mut average: ResMut<Average>,
    mut reader: EventReader<WebcamFacialDataEvent>,
) {
    for event in reader.iter() {
        let x = event.0.center_x as f32;
        let y = event.0.center_y as f32;
        average.x = (x + average.x) / 2.0;
        average.y = (y + average.y) / 2.0;
        for mut head in head.iter_mut() {
            head.translation += 0.005;
        }
    }
}


examples/object_mover.rs
use bevy::pbr::DirectionalLightShadowMap;
use bevy::prelude::*;

use bevy_webcam_facial::*;

#[derive(Component)]
struct WebcamControlledObject;

fn main() {
    App::new()
        .insert_resource(DirectionalLightShadowMap { size: 2048 })
        .add_plugins(DefaultPlugins)
        // Add plugin with a *MUST* camera parameters
        .add_plugins(WebcamFacialPlugin::default())
        .add_systems(Startup, setup)
        // Add system to read data events and do something with data
        .add_systems(Update, move_object)
        .run();
}

/// set up a simple 3D scene
fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    // Ground plane
    commands.spawn(PbrBundle {
        mesh: meshes.add(shape::Plane::from_size(5.0).into()),
        material: materials.add(Color::rgb(0.3, 0.5, 0.3).into()),
        ..default()
    });
    // Cube object with WebcamControlledObject
    commands.spawn((
        PbrBundle {
            mesh: meshes.add(Mesh::from(shape::Cube { size: 1.0 })),
            material: materials.add(Color::rgb(0.8, 0.7, 0.6).into()),
            transform: Transform::from_xyz(0.0, 1.0, 0.0),
            ..default()
        },
        // Mark object
        WebcamControlledObject,
    ));
    // Some light
    commands.spawn(PointLightBundle {
        point_light: PointLight {
            intensity: 500.0,
            shadows_enabled: true,
            ..default()
        },
        transform: Transform::from_xyz(0.0, 4.0, 0.0),
        ..default()
    });
    // Camera for scene
    commands.spawn(Camera3dBundle {
        transform: Transform::from_xyz(0.0, 3.0, 4.0).looking_at(Vec3::ZERO, Vec3::Y),
        ..default()
    });
}

fn move_object(
    // Get object to control with WebcamControlledObject
    mut query: Query<&mut Transform, With<WebcamControlledObject>>,
    // Read data event with camera data
    mut reader: EventReader<WebcamFacialDataEvent>,
) {
    for event in reader.iter() {
        // Get data from event.
        let x = event.0.center_x;
        let y = event.0.center_y;
        // Print coords and do basic transforms
        info!("{:?} {:?}", x, y);
        for mut transform in query.iter_mut() {
            // Move object with x10 less influence, also add 2 to z transform
            transform.translation.x = x as f32 / 10.0;
            transform.translation.z = y as f32 / 10.0 + 2.0;
        }
    }
}


examples/kalman filter.rs
use kalman_rs::{KalmanFilter, MeasurementModel, TransitionModel};
use nalgebra::{Matrix2, Vector2};

fn main() {
    let transition_model = TransitionModel::new(Matrix2::new(1.0, 0.0, 0.0, 1.0)); // Assuming a constant velocity model in 2D (x and y are independent)
    let measurement_model = MeasurementModel::new(Matrix2::identity()); // Assuming we can directly measure position

    // Adjust the process noise covariance matrix to increase the uncertainty in the transition model
    let process_noise_covariance = Matrix2::new(0.01, 0.0, 0.0, 0.01);
    let mut kalman_filter = KalmanFilter::new_with_process_noise(
        &transition_model,
        &measurement_model,
        &process_noise_covariance,
    );

    // Simulate receiving noisy data every 10ms
    loop {
        let noisy_coordinate = get_noisy_coordinate(); // Replace this with your real data input

        // Perform Kalman filter prediction step (assuming constant velocity model)
        kalman_filter.predict();

        // Adjust the measurement noise covariance matrix to increase the influence of noisy measurements
        let measurement_noise_covariance = Matrix2::new(0.01, 0.0, 0.0, 0.01);
        kalman_filter.update_with_measurement_noise(
            &Vector2::new(noisy_coordinate.x, noisy_coordinate.y),
            &measurement_noise_covariance,
        );

        // Get the smoothed estimate from the Kalman filter
        let smoothed_state = kalman_filter.state_estimate();

        // Use the smoothed_state as needed (e.g., visualization or further processing)
        let smoothed_coordinate = Coordinate::new(smoothed_state[0], smoothed_state[1]);
    }
}


examples/camera_control.rs
use bevy::app::AppExit;
use bevy::pbr::DirectionalLightShadowMap;
use bevy::prelude::*;
use bevy_scene_hook::{HookPlugin, HookedSceneBundle, SceneHook};

use bevy_webcam_facial::*;

#[derive(Component)]
struct CameraControl;

#[derive(Component)]
struct CameraEye;

fn main() {
    App::new()
        .insert_resource(DirectionalLightShadowMap { size: 2048 })
        .add_plugins(DefaultPlugins)
        .add_plugins(WebcamFacialPlugin {
            config_webcam_device: 0,
            config_webcam_width: 640,
            config_webcam_height: 480,
            config_webcam_framerate: 33,
            config_webcam_autostart: true,
            // Using LowPass filter, with value of 'alpha' at 0.01 for last 20 frames to get more smoothing
            config_filter_length: 20,
            config_filter_type: SmoothingFilterType::LowPass(0.01),
        })
        // Using HookPlugin to get named object from loaded gltf scene
        .add_plugins(HookPlugin)
        .add_systems(Startup, load_scene)
        .add_systems(
            Update,
            (
                set_camera_position_from_plugin,
                user_input_to_plugin_control_system,
            ),
        )
        .run();
}

/// set up 3D scene
fn load_scene(mut commands: Commands, asset_server: Res<AssetServer>) {
    commands.spawn(HookedSceneBundle {
        scene: SceneBundle {
            scene: asset_server.load("rooster.gltf#Scene0"),
            ..default()
        },
        hook: SceneHook::new(|entity, cmds| {
            match entity.get::<Name>().map(|t| t.as_str()) {
                Some("Camera") => {
                    cmds.insert(CameraEye);
                }
                _ => {}
            };
        }),
    });
    commands.spawn(PointLightBundle {
        point_light: PointLight {
            intensity: 500.0,
            shadows_enabled: true,
            ..default()
        },
        transform: Transform::from_xyz(0.0, 4.0, 0.0),
        ..default()
    });
    commands.spawn(
        TextBundle::from_section(
            "A: Start capture\nS: Stop capture\nESC: Exit app",
            TextStyle {
                font_size: 20.0,
                color: Color::BLUE,
                ..default()
            },
        )
        .with_style(Style {
            position_type: PositionType::Absolute,
            top: Val::Px(5.0),
            left: Val::Px(15.0),
            ..default()
        }),
    );
}

fn set_camera_position_from_plugin(
    mut camera: Query<&mut Transform, With<CameraEye>>,
    mut webcam_data: EventReader<WebcamFacialDataEvent>,
    _time: Res<Time>,
) {
    for event in webcam_data.iter() {
        let x = event.0.center_x;
        let y = -event.0.center_y + 3.0;

        for mut transform in camera.iter_mut() {
            // For camera moving while looking at target:
            transform.translation = Transform::from_xyz(x, y, 10.0)
                .looking_at(
                    Vec3 {
                        x: 0.0,
                        y: 1.0,
                        z: 0.0,
                    },
                    Vec3::Y,
                )
                .translation;
            transform.rotation = Transform::from_xyz(x, y, 10.0)
                .looking_at(
                    Vec3 {
                        x: 0.0,
                        y: 1.0,
                        z: 0.0,
                    },
                    Vec3::Y,
                )
                .rotation;

            /* OR */

            // For FPS Mode (there is a better way to do this using bevy_fly_camera or bevy_orbit_controls or something else, plus easing/interpolation...):
            // let pitch = y - 3.0;
            // let yaw = -x;
            // transform.rotation = Quat::from_axis_angle(Vec3::Y, yaw.to_radians() * 10.0)
            //     * Quat::from_axis_angle(-Vec3::X, pitch.to_radians() * 10.0);
        }
    }
}

// Keyboard control for plugin stop/start and app
fn user_input_to_plugin_control_system(
    keyboard_input: Res<Input<KeyCode>>,
    mut webcam_facial_control: ResMut<WebcamFacialController>,
    mut exit: EventWriter<AppExit>,
) {
    if keyboard_input.just_pressed(KeyCode::A) {
        webcam_facial_control.control = true;
    }
    if keyboard_input.just_pressed(KeyCode::S) {
        webcam_facial_control.control = false;
    }
    if keyboard_input.just_pressed(KeyCode::Escape) {
        exit.send(AppExit);
    }
}


src/lib.rs
// Plugin that reads webcamera, detects face calculates frame box
// and sends coordinates to Bevy as Event.

use bevy::{
    app::{App, Plugin, Update},
    ecs::{
        component::Component,
        entity::Entity,
        event::{Event, EventWriter},
        system::{Commands, Query, ResMut, Resource},
    },
    log::{debug, error, info},
    tasks::{AsyncComputeTaskPool, Task},
};

use crossbeam_channel::{bounded, Receiver, SendError, Sender};
use futures_lite::future;
use std::sync::{
    atomic::{AtomicBool, Ordering},
    Arc,
};

// camera capture
use camera_capture;
// rustface detector
use rustface::ImageData;
// image utils
use image::{ImageBuffer, Luma};
// Data filter/smoothing
mod filter;
use filter::WebcamFacialDataFiltered;
pub use filter::SmoothingFilterType;

pub struct WebcamFacialPlugin {
    pub config_webcam_device: u32,
    pub config_webcam_width: u32,
    pub config_webcam_height: u32,
    pub config_webcam_framerate: u32,
    pub config_webcam_autostart: bool,
    pub config_filter_type: SmoothingFilterType,
    pub config_filter_length: u32,
}
// Plugin configuration for webcam to be accesible from plugin system
#[derive(Resource)]
pub struct WebcamFacialController {
    pub sender: Sender<WebcamFacialData>,
    pub receiver: Receiver<WebcamFacialData>,
    pub control: bool,
    pub status: Arc<AtomicBool>,
    config_device: u32,
    config_width: u32,
    config_height: u32,
    config_framerate: u32,
    config_filter_type: SmoothingFilterType,
    config_filter_length: u32,
}

#[derive(Component)]
struct WebcamFacialTask(Task<bool>);

// WebcamFacialEvent event for sending WebcamFacialData to main Bevy app
#[derive(Event)]
pub struct WebcamFacialDataEvent(pub WebcamFacialData);

// Data structure to be exchanged with Bevy
#[derive(Default, Clone, Debug)]
pub struct WebcamFacialData {
    pub center_x: f32,
    pub center_y: f32,
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
    pub score: f32,
}

impl Plugin for WebcamFacialPlugin {
    fn build(&self, app: &mut App) {
        // Add thread channels for data exchange
        let (task_channel_sender, task_channel_receiver) = bounded(1);
        let task_status = Arc::new(AtomicBool::new(false));
        // Store plugin control,data channels and settings in a resource
        let plugin = WebcamFacialController {
            sender: task_channel_sender,
            receiver: task_channel_receiver,
            control: self.config_webcam_autostart.clone(),
            status: task_status,

            config_device: self.config_webcam_device.clone(),
            config_width: self.config_webcam_width.clone(),
            config_height: self.config_webcam_height.clone(),
            config_framerate: self.config_webcam_framerate.clone(),
            config_filter_type: self.config_filter_type.clone(),
            config_filter_length: self.config_filter_length.clone(),
        };
        // Insert nesecary resources, events and systems
        app.insert_resource(plugin)
            .add_event::<WebcamFacialDataEvent>()
            .add_systems(Update, webcam_facial_task_runner);
    }
}

impl Default for WebcamFacialPlugin {
    fn default() -> Self {
        Self {
            config_webcam_device: 0,
            config_webcam_width: 640,
            config_webcam_height: 480,
            config_webcam_framerate: 15,
            config_webcam_autostart: true,
            config_filter_type: SmoothingFilterType::LowPass(0.1),
            config_filter_length: 10,
        }
    }
}

fn webcam_facial_task_runner(
    mut webcam_facial: ResMut<WebcamFacialController>,
    mut commands: Commands,
    mut plugin_task: Query<(Entity, &mut WebcamFacialTask)>,
    mut plugin_events: EventWriter<WebcamFacialDataEvent>,
) {
    // If enabled and not running - start task
    if webcam_facial.control & !webcam_facial.status.load(Ordering::SeqCst) {
        // Get Arc clones
        let task_running = webcam_facial.status.clone();
        let sender_clone = webcam_facial.sender.clone();

        let camera_device = webcam_facial.config_device;
        let camera_width = webcam_facial.config_width;
        let camera_height = webcam_facial.config_height;
        let camera_framerate = webcam_facial.config_framerate;
        let filter_type = webcam_facial.config_filter_type;
        let filter_length = webcam_facial.config_filter_length;

        info!("Starting plugin");
        let thread_pool = AsyncComputeTaskPool::get();
        // Main task and its loop
        let task = thread_pool.spawn(async move {
            // Initialize webcam
            let mut cam_iter = match get_camera_frame_iterator(
                camera_device,
                camera_width,
                camera_height,
                camera_framerate,
            ) {
                Some(iter) => iter,
                None => {
                    return false;
                }
            };
            // Initialize face detector
            //TODO Model selection, remove hardcoded
            let mut detector =
                match rustface::create_detector(&"assets/NN_Models/seeta.bin".to_string()) {
                    Ok(mut detector) => {
                        info!("Using assets/NN_Models/seeta.bin recognition model.");
                        detector.set_min_face_size(20);
                        detector.set_score_thresh(2.0);
                        detector.set_pyramid_scale_factor(0.8);
                        detector.set_slide_window_step(4, 4);
                        detector
                    }
                    Err(error) => {
                        error!("Failed to create detector: {}", error.to_string());
                        return false;
                    }
                };

            let mut filtered_data = WebcamFacialDataFiltered::new(filter_length, filter_type);
            info!("Capturing frames...");
            while task_running.load(Ordering::SeqCst) {
                // Get frame from buffer
                let rgb_frame = cam_iter.next().unwrap();
                // Convert RGB frame to grayscale
                let grayscale_image = ImageBuffer::from_fn(camera_width, camera_height, |x, y| {
                    let rgb_pixel = *rgb_frame.get_pixel(x, y);
                    let gray_value = rgb_pixel[0] as u32 * 77
                        + rgb_pixel[1] as u32 * 150
                        + rgb_pixel[2] as u32 * 29;
                    Luma([((gray_value >> 8) & 0xFF) as u8])
                });
                // Get Image data from buffer data
                let grayscale_image_data =
                    ImageData::new(&grayscale_image, camera_width, camera_height);

                // Detect face data in provided image data
                let faces = detector.detect(&grayscale_image_data);

                // Initialize zero values if face not found
                let mut facial_data = WebcamFacialData::default();

                // Get face with maximum human face probability (best candidate)
                let max_face = faces.iter().max_by_key(|p| p.score() as i32);
                match max_face {
                    Some(max_face) => {
                        debug!("Max score face: {:?}", max_face);
                        // Take face rectangle coords and score
                        facial_data.x = faces[0].bbox().x() as f32;
                        facial_data.y = faces[0].bbox().y() as f32;
                        facial_data.width = faces[0].bbox().width() as f32;
                        facial_data.height = faces[0].bbox().height() as f32;
                        facial_data.score = faces[0].score() as f32;

                        // Calculate the scale factor to map the camera resolution
                        let w_scale_factor = 100.0 / camera_width as f32;
                        let h_scale_factor = 100.0 / camera_width as f32;

                        // Calculate the coordinates and dimensions in the desired range (-50.0) to (50.0)
                        facial_data.x = facial_data.x * w_scale_factor - 50.0;
                        facial_data.y = facial_data.y * h_scale_factor - 50.0;
                        facial_data.width = facial_data.width * w_scale_factor;
                        facial_data.height = facial_data.height as f32 * h_scale_factor;
                        facial_data.center_x = (2.0 * facial_data.x + facial_data.width) / -2.0; // minus flips values so negative is left
                        facial_data.center_y = (2.0 * facial_data.y + facial_data.height) / 2.0;
                    }
                    None => {
                        debug!("No faces found. Using default zero values.");
                    }
                }
                filtered_data.push(facial_data);

                // Send processed and filtered data
                match sender_clone.send(filtered_data.get()) {
                    Ok(()) => {
                        debug!("Data from task sent.")
                    }
                    Err(SendError(data)) => {
                        error!("Failed to send task data: {:?}", data);
                    }
                }
            }
            true
        });
        commands.spawn(WebcamFacialTask(task));
        // Set flag that we started thread
        webcam_facial.status.store(true, Ordering::SeqCst);
    }
    // If not enabled and task is running set flag to stop
    if !webcam_facial.control & webcam_facial.status.load(Ordering::SeqCst) {
        webcam_facial.status.store(false, Ordering::SeqCst);
    }
    for (entity, mut task) in &mut plugin_task {
        if let Some(status) = future::block_on(future::poll_once(&mut task.0)) {
            // Task completed, so remove task component from entity
            commands.entity(entity).remove::<WebcamFacialTask>();
            webcam_facial.status.store(false, Ordering::SeqCst);
            webcam_facial.control = false;
            if status {
                info!("Camera stopped.");
            } else {
                info!("Plugin setup failed. Plugin self disabled.");
            }
        }
    }
    while let Ok(data) = webcam_facial.receiver.try_recv() {
        debug!("Send Bevy event {:?}", data);
        plugin_events.send(WebcamFacialDataEvent(data));
    }
}

fn get_camera_frame_iterator(
    camera_device: u32,
    camera_width: u32,
    camera_height: u32,
    camera_framerate: u32,
) -> Option<camera_capture::ImageIterator> {
    // Create the camera device
    let camera_device = match camera_capture::create(camera_device) {
        Ok(device) => {
            #[cfg(unix)]
            info!("Using '/dev/video{}' camera.", camera_device);
            #[cfg(windows)]
            info!("Using camera ID:{}.", camera_device);
            device
        }
        Err(err) => {
            error!(
                "Error creating camera device [{}]: {:?}",
                camera_device, err
            );
            return None;
        }
    };
    // Set the resolution
    let resolution_device = match camera_device.resolution(camera_width, camera_height) {
        Ok(resolution) => {
            info!(
                "Camera resolution set to {}x{}.",
                camera_width, camera_height
            );
            resolution
        }
        Err(err) => {
            error!("Error setting camera resolution: {:?}", err);
            return None;
        }
    };
    // Set the frame rate and start the camera capture
    let cam_iter = match resolution_device.fps(camera_framerate as f64) {
        Ok(fps) => {
            info!("Camera fps set to {}.", camera_framerate);
            fps.start()
        }
        Err(err) => {
            error!("Error setting camera frame rate: {:?}", err);
            return None;
        }
    };
    match cam_iter {
        Ok(iter) => Some(iter),
        Err(err) => {
            error!("Error starting camera: {:?}", err);
            None
        }
    }
}
